db/migrate/20220308201942_create_upload_references.rb

```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class CreateUploadReferences < ActiveRecord::Migration[6.1]
+  def change
+    create_table :upload_references do |t|
+      t.references :upload, null: false
+      t.references :target, polymorphic: true, null: false
+      t.timestamps
+    end
+
+    add_index :upload_references, [:upload_id, :target_type, :target_id], unique: true, name: 'index_upload_references_on_upload_and_target'
```

Comment:
I am not sure if this will work, I am 95% sure that currently an upload can be linked to multiple posts, which has been an issue with secure media things. I am not sure if this is the case with other upload references, but yeah I don't think UNIQUE will work for this.